### PR TITLE
Use control flow AD in layers and add leak checking utility.

### LINKED ISSUE #389: Remove all unnecessary custom VJPs for differentiable control flow.
https://github.com/apple/swift/pull/26355 fixed all control flow memory leaks (#364). After that patch lands, we should remove all unnecessary custom VJPs for differentiable control flow, such as the custom VJP for `BatchNorm.callAsFunction(_:)`.

### PR DESCRIPTION
**Pull request by rxwei**
Remove custom VJPs from all layers so that they go through control flow differentiation. This applies to `BatchNorm` and `Dropout`. `Dense` was already using control flow differentiation.

Introduce a tensor leak checking test utility `withTensorLeakChecking { ... }` and use it in tests. A `globalTensorCount` variable is added to the thread-local `Context` to keep track of the number of live tensors, and `TFETensorHandle`'s `init(_owning:)` and `deinit` increases and decreases `globalTensorCount`, respectively.

Resolves #389.

### CODE REVIEW

**Comment by rxwei** [comment]
Tests are failing because the toolchain is old. We are updating it later today.

File: Tests/TensorFlowTests/Helpers.swift [hunk 1]
```diff
@@ -36,3 +36,13 @@ internal func assertEqual<T: TensorFlowFloatingPoint>(
 ) {
     assertEqual(x.scalars, y.scalars, accuracy: accuracy, message, file: file, line: line)
 }
+
+func withTensorLeakChecking(file: StaticString = #file,
```
Comment (eaplatanios) [inline Tests/TensorFlowTests/Helpers.swift]: Should we follow the convention of formatting long argument lists as follows? It's pretty much done like that in most of the `swift-apis` repo.

```swift
func withTensorLeakChecking(
    file: StaticString = #file,
    line: UInt = #line,
    _ body: () throws -> Void
) rethrows {
```

File: Tests/TensorFlowTests/LayerTests.swift [hunk 2]
```diff
@@ -43,14 +43,18 @@ final class LayerTests: XCTestCase {
             let optimizer = SGD(for: model)
             let x = Tensor<Float>([[0, 0], [0, 1], [1, 0], [1, 1]])
             let y = Tensor<Float>([0, 1, 1, 0])
-            let initialLoss = meanSquaredError(predicted: model(x).squeezingShape(at: 1), expected: y)
-            for _ in 0..<10 {
-                let 𝛁model = model.gradient { model -> Tensor<Float> in
-                    meanSquaredError(predicted: model(x).squeezingShape(at: 1), expected: y)
+            let initialLoss = meanSquaredError(predicted: model(x).squeezingShape(at: 1),
```
Comment (eaplatanios) [inline Tests/TensorFlowTests/LayerTests.swift]: How about this?
```swift
let initialLoss = meanSquaredError(
    predicted: model(x).squeezingShape(at: 1),
    expected: y)
```
I guess we should decide on a style and document it somewhere.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 3]
```diff
@@ -43,14 +43,18 @@ final class LayerTests: XCTestCase {
...
+                for _ in 0..<10 {
+                    let 𝛁model = model.gradient { model -> Tensor<Float> in
+                        meanSquaredError(predicted: model(x).squeezingShape(at: 1), expected: y)
+                    }
+                    optimizer.update(&model, along: 𝛁model)
                 }
-                optimizer.update(&model, along: 𝛁model)
             }
-            let updatedLoss = meanSquaredError(predicted: model(x).squeezingShape(at: 1), expected: y)
+            let updatedLoss = meanSquaredError(predicted: model(x).squeezingShape(at: 1),
```
Comment (eaplatanios) [inline Tests/TensorFlowTests/LayerTests.swift]:
```swift
let updatedLoss = meanSquaredError(
    predicted: model(x).squeezingShape(at: 1),
    expected: y)
```

File: Tests/TensorFlowTests/LayerTests.swift [hunk 4]
```diff
@@ -298,7 +303,8 @@ final class LayerTests: XCTestCase {
       let layer = UpSampling3D<Float>(size: size)
       let input = Tensor<Float>(shape: [1, 4, 3, 2, 1], scalars: (0..<24).map(Float.init))
       let output = layer.inferring(from: input)
-      let expected = TensorShape([1, input.shape[1] * size, input.shape[2] * size, input.shape[3] * size, 1])
+      let expected = TensorShape([1, input.shape[1] * size, input.shape[2] * size,
```
Comment (eaplatanios) [inline Tests/TensorFlowTests/LayerTests.swift]:
```swift
      let expected = TensorShape(
          [1, input.shape[1] * size, input.shape[2] * size, input.shape[3] * size, 1])
```

File: Tests/TensorFlowTests/Helpers.swift (hunk 1 above)
Comment (eaplatanios) [reply Tests/TensorFlowTests/Helpers.swift]: +1 for this helper and the memory leak tests btw :)

File: Tests/TensorFlowTests/LayerTests.swift (hunk 2 above)
Comment (rxwei) [reply Tests/TensorFlowTests/LayerTests.swift]: Done. I lean towards Google Swift Style Guide's recommendations although I don't always remember to follow it.

**Comment by lattner** [comment]
This is awesome!
